src/cmd/sanity_test/sanity_tests.go

```diff
@@ -105,260 +123,333 @@ func checkEmailRestoration(
 	}
 
-	for _, restoreData := range restoreResponses.GetValue() {
-		restoreName := *restoreData.GetName()
-
-		if restoreData.GetFile() != nil {
-			if *restoreData.GetSize() != file[restoreName] {
-				fmt.Printf("Size of file %s is different in drive %d and restored file: %d ",
-					restoreName,
-					file[restoreName],
-					*restoreData.GetSize())
+	for _, item := range restored.GetValue() {
+		var (
+			itemID   = ptr.Val(item.GetId())
+			itemName = ptr.Val(item.GetName())
+			itemSize = ptr.Val(item.GetSize())
+		)
+
+		if item.GetFile() != nil {
+			if itemSize != fileSizes[itemName] {
+				fmt.Println("File size does not match:")
+				fmt.Println("*  expected:", fileSizes[itemName])
+				fmt.Println("*  actual:", itemSize)
+				fmt.Println("Item:", itemName, itemID)
 				os.Exit(1)
 			}
 
 			continue
 		}
 
-		itemBuilder := client.DrivesById(driveID).ItemsById(*restoreData.GetId())
-
-		if restoreData.GetFolder() != nil {
-			permissionColl, err := itemBuilder.Permissions().Get(context.TODO(), nil)
-			if err != nil {
-				fmt.Printf("Error getting permission: %v\n", err)
-				os.Exit(1)
-			}
+		if item.GetFolder() == nil && item.GetPackage() == nil {
+			continue
+		}
 
-			userPermission := []string{}
+		var (
+			expectItem = folderPermission[itemID]
+			results    = permissionsIn(ctx, client, driveID, itemID, nil)
+		)
 
-			for _, perm := range permissionColl.GetValue() {
-				userPermission = perm.GetRoles()
-			}
+		for pid, result := range results {
+			expect := expectItem[pid]
 
-			if !reflect.DeepEqual(folderPermission[restoreName], userPermission) {
-				fmt.Printf("Permission mismatch for %s.", restoreName)
+			if !slices.Equal(expect, result) {
+				fmt.Println("permissions are not equal")
+				fmt.Println("*  expected: ", expect)
+				fmt.Println("*  actual: ", result)
+				fmt.Println("Item:", itemName, itemID)
+				fmt.Println("Permission:", pid)
 				os.Exit(1)
 			}
 		}
 	}
 }
+
+// ---------------------------------------------------------------------------
+// Helpers
+// ---------------------------------------------------------------------------
+
+func fatal(ctx context.Context, msg string, err error) {
+	logger.CtxErr(ctx, err).Error("test failure: " + msg)
+	fmt.Println(msg+": ", err)
+	os.Exit(1)
+}
+
+func permissionsIn(
+	ctx context.Context,
+	client *msgraphsdk.GraphServiceClient,
+	driveID, itemID string,
+	init map[string][]string,
+) map[string][]string {
+	result := map[string][]string{}
+
+	pcr, err := client.
+		DrivesById(driveID).
+		ItemsById(itemID).
+		Permissions().
+		Get(ctx, nil)
+	if err != nil {
+		fatal(ctx, "getting permission", err)
+	}
+
+	if len(init) > 0 {
+		maps.Copy(result, init)
+	}
+
+	for _, p := range pcr.GetValue() {
+		var (
+			pid   = ptr.Val(p.GetId())
+			roles = p.GetRoles()
+		)
+
+		slices.Sort(roles)
+
+		result[pid] = roles
+	}
+
+	return result
+}
+
+func mustGetTimeFromName(ctx context.Context, name string) time.Time {
```

Comment:
maybe we should just turn this into a function that returns a bool as to whether we should skip the folder or not? That would remove the remaining if-condition in the calling code, contain error handling to this, and make reading the other code a bit easier maybe?

Reply:
yeah, not a bad idea.  Obviously my brain is only half into this.  Thanks for the checks